File: main/tests/UnitTests/MonoDevelop.CSharpBinding/TestWorkbenchWindow.cs
//
// TestWorkbenchWindow.cs
//
// Author:
//   Mike Krüger <mkrueger@novell.com>
//
// Copyright (C) 2008 Novell, Inc (http://www.novell.com)
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

using System;
using System.Collections;

using MonoDevelop.Ide.Gui;
using System.Collections.Generic;
using Mono.Addins;

namespace MonoDevelop.CSharpBinding
{
	public class TestWorkbenchWindow : IWorkbenchWindow
	{
		string title;
		public string Title {
			get {
				return title;
			}
			set {
				title = value;
			}
		}
		
		public string DocumentType {
			get { return ""; }
			set {}
		}
		
		public Document Document {
			get;
			set;
		}

		public bool ShowNotification {
			get { return false; }
			set {}
		}
		
		IViewContent viewContent;
		public IViewContent ViewContent {
			get { return viewContent; }
			set { viewContent = value; }
		}

		public ExtensionContext ExtensionContext {
			get { return AddinManager.AddinEngine; }
		}
		
		public IEnumerable<IAttachableViewContent> SubViewContents { get { return new IAttachableViewContent[0]; } }

		public IBaseViewContent ActiveViewContent {
			get { return ViewContent;}
			set {}
		}
		
		public bool CloseWindow (bool force)
		{
			return true;
		}
		
		public void SelectWindow ()
		{
		}
		
		public void SwitchView (int viewNumber)
		{
		}
		public void SwitchView (IAttachableViewContent view)
		{
		}

		public void SwitchToSingleMode ()
		{
		}
		
		public int FindView<T> ()
		{
			return -1;
		}
		
		public void AttachViewContent (IAttachableViewContent subViewContent)
		{
			
		}

		public void InsertViewContent (int index, IAttachableViewContent subViewContent)
		{
			
		}

		public DocumentToolbar GetToolbar (IBaseViewContent targetView)
		{
			throw new NotImplementedException ();
		}

		public event EventHandler DocumentChanged;
		public event WorkbenchWindowEventHandler Closing;
		public event WorkbenchWindowEventHandler Closed;
		public event ActiveViewContentEventHandler ActiveViewContentChanged;
		public event EventHandler ViewsChanged;

		DocumentToolbar IWorkbenchWindow.GetToolbar (IBaseViewContent targetView)
		{
			throw new System.NotImplementedException ();
		}
	}
}

File: monodevelop/MonoDevelop.FSharpBinding/MonoDevelop.FSharp.Tests/TestWorkbenchWindow.fs
namespace MonoDevelopTests
open System
open Mono.Addins
open MonoDevelop.Ide.Gui

type TestWorkbenchWindow() =
    let mutable viewContent = Unchecked.defaultof<IViewContent>
    let edc = DelegateEvent<_>()
    let closed = DelegateEvent<_>()
    let closing = DelegateEvent<_>()
    let avcc = DelegateEvent<_>()
    let viewsChanged = DelegateEvent<_>()

    member x.SetViewContent(v) = viewContent <- v
    
    interface IWorkbenchWindow with
        member x.SelectWindow () = ()
        member x.SwitchView (viewNumber:int) = ()
        member x.SwitchView (viewNumber:IAttachableViewContent) = ()
        member x.FindView<'a>() = -1
        member val Title = "" with get,set
        member val Document = null with get, set
        member val DocumentType = "" with get, set
        member val ShowNotification = false with get, set
        member x.ViewContent with get() = viewContent
        member x.ActiveViewContent with get() = viewContent :> _ and set v = viewContent <- downcast v
        member x.ExtensionContext with get() = AddinManager.AddinEngine :> _
        member x.CloseWindow (force) = true
        member x.AttachViewContent (subViewContent) = ()
        member x.InsertViewContent(index, subViewContent) = ()
        member x.SubViewContents with get() = Seq.empty
        member x.GetToolbar(targetView) = failwith "Not Implemented"

        [<CLIEvent>]
        member x.DocumentChanged = edc.Publish
        [<CLIEvent>]
        member x.Closed = closed.Publish
        [<CLIEvent>]
        member x.Closing = closing.Publish
        [<CLIEvent>]
        member x.ActiveViewContentChanged = avcc.Publish
        [<CLIEvent>]
        member x.ViewsChanged = viewsChanged.Publish